hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -579,42 +925,67 @@ void GlobalOptimizer::run() {
         return std::vector<NodeRef>{units.begin(), units.end()};
     }();
 
+    const auto passes__ = rt::getenv("HILTI_OPTIMIZER_PASSES");
+    const auto passes_ =
+        passes__ ? std::optional(util::split(*passes__, ":")) : std::optional<std::vector<std::string>>();
+    auto passes = passes_ ? std::optional(std::set<std::string>(passes_->begin(), passes_->end())) :
+                            std::optional<std::set<std::string>>();
+
+    if ( ! passes || passes->count("feature_requirements") ) {
+        // The `FeatureRequirementsVisitor` enables or disables code
+        // paths and needs to be run before all other passes since
+        // it needs to see the code before any optimization edits.
+        FeatureRequirementsVisitor v;
+        for ( auto& unit : units )
+            v.collect(*unit);
+
+        for ( auto& unit : units )
+            v.transform(*unit);
+    }
+
+    const std::map<std::string, std::function<std::unique_ptr<OptimizerVisitor>()>> creators =
+        {{"constant_folding", []() { return std::make_unique<ConstantFoldingVisitor>(); }},
+         {"functions", []() { return std::make_unique<FunctionVisitor>(); }},
+         {"types", []() { return std::make_unique<TypeVisitor>(); }}};
+
+    // If no user-specified passes are given enable all of them.
+    if ( ! passes ) {
+        passes = std::set<std::string>();
+        for ( const auto& [pass, _] : creators )
+            passes->insert(pass);
+    }
+
     while ( true ) {
         bool modified = false;
 
-        FunctionVisitor function_visitor;
-        TypeVisitor type_visitor;
+        // NOTE: We do not use `util::transform` here to guarantee a consisted order of the visitors.
+        std::vector<std::unique_ptr<OptimizerVisitor>> vs;
+        vs.reserve(passes->size());
+        for ( const auto& pass : *passes )
+            vs.push_back(creators.at(pass)());
```

Comment:
As `passes` might come from the user we need to check whether a certain `pass` is known to `creators` here.